### PR TITLE
Update pox_4_tests to run in 2.5 and 3.0

### LINKED ISSUE #5049: [Nakamoto] Investigate how reward cycle is calculated from burn block height
#5037 revealed that some of the miner code had an off-by-one error when calculating the reward cycle index, and thus fetching a reward set, for a given burn block height. For example, with 20 block cycles, the miner would think that burn height 240 would correspond to cycle 11, when it should be 12.

The main function in question is `burnchain.pox_reward_cycle(burn_election_height)`, which uses a different calculation than `block_height_to_reward_cycle`.

This ticket is mainly open-ended, as we don't know exactly where (or if) there are other parts in the code that should be changed. Instead, this ticket refers to investigation and then potential changes.

### PR DESCRIPTION
**Pull request by hstove**
- Closes #5049

To help catch edge cases related to Nakamoto, this PR updates the `pox_4_tests` test suite to run tests in both epoch 2.5 and 3.0. Previously, all tests only ran in 2.5.

This PR also includes the work from @kantai to add consistent behavior to mod 0 blocks in Nakamoto, which was the cause of a previous reward cycle calculation bug. That work is included in https://github.com/stacks-network/stacks-core/pull/5072/commits/1d836175730eff88eaec9440c608267eb6303494.

This PR adds a `nakamoto_cases` rstest template, so that tests will run twice, and they're provided a `use_nakamoto` parameter. Certain setup functions have been updated (such as `prepare_pox4_test`) to utilize this argument. When `use_nakamoto` is true, the setup function utilizes the `NakamotoBootPlan` struct to boot into Nakamoto. Using `NakamotoBootPlan` adds things like automatically stacking a single signer and uses the Nakamoto VRF proof when mining blocks.

For the most part, the tests themselves are not updated. There are certain instances, such as expecting a reward set with 1 signer, which are updated to expect 2 signers when in epoch 3.0.

I've also added a helper function, `tenure_with_txs`, which mimicks the `peer.tenure_with_txs` function. In 3.0, this function creates a tenure with a Nakamoto block that includes the transactions.

### CODE REVIEW

**Comment by hstove** [comment]
Ok @obycode and @jcnelson , this is ready for review. There are really two separate things to review here:

- Aaron's commit that makes reward cycle calculation more consistent, which is pretty consensus critical
- All the pox-4 test updates, which were done to help catch any remaining reward cycle calculation bugs

**Review by obycode — APPROVED** [review]
This looks good. I just had one comment about a remaining todo.

**Review by jcnelson — COMMENTED** [review]
A couple things:

* I think you forgot to update `getpoxinfo.rs`.  It always uses `reward_cycle_to_block_height()`, but should probably use `nakamoto_first_block_of_cycle` if the current burnchain height is in epoch 3.0.

* Similarly, I see that `getstackers.rs` continues to use `reward_cycle_to_block_height` in epoch 3.0.

* I am going to need to make some changes to the Nakamoto inventory sync state machine and downloader, which assume that reward cycles start at reward index 1, not 0.  But I'll do that in my "more miner fixes" PR, into which I'll merge this PR as well.

**Comment by hstove** [comment]
@jcnelson regarding your points:

> I think you forgot to update getpoxinfo.rs. It always uses reward_cycle_to_block_height(), but should probably use nakamoto_first_block_of_cycle if the current burnchain height is in epoch 3.0.

`reward_cycle_to_block_height` considers the modulo 0 block to be the start of the reward cycle, so I think this is already consistent. `nakamoto_first_block_of_cycle`, is the "reverse" operation, but it also uses the modulo 0 block height.

> Similarly, I see that getstackers.rs continues to use reward_cycle_to_block_height in epoch 3.0

I'm a little unsure of how this should be handled. This function is only used to verify that pox-4 is the current active PoX contract, and it returns an error otherwise. From 2.5 onwards, this will simply pass.

If the current burn height is 40, and 2.5 activates at 41, what is the expected behavior here (assuming 20 block cycles)? Should it error?

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 1]
```diff
@@ -230,7 +256,7 @@ fn pox_extend_transition() {
     //  tenures start being tracked.
     let EMPTY_SORTITIONS = 25;
 
-    let (epochs, pox_constants) = make_test_epochs_pox();
+    let (epochs, pox_constants) = make_test_epochs_pox(false);
```
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:259]: I'm curious why we aren't considering a case of this test with `make_test_epochs_pox(true)`?  Same question for those below.  Is it because these tests require more than 123 sortitions?

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 2]
```diff
@@ -5209,25 +5266,39 @@ fn stack_stx_signer_key() {
...
     let reward_cycle_ht = burnchain.reward_cycle_to_block_height(next_reward_cycle);
-    let mut reward_set = get_reward_set_entries_at(&mut peer, &latest_block, reward_cycle_ht);
-    assert_eq!(reward_set.len(), 1);
-    let reward_entry = reward_set.pop().unwrap();
-    assert_eq!(
-        PoxAddress::try_from_pox_tuple(false, &pox_addr_val).unwrap(),
-        reward_entry.reward_address
-    );
+    let reward_set = get_reward_set_entries_at(&mut peer, &latest_block, reward_cycle_ht);
+    assert_eq!(reward_set.len(), {
```
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:5270]: Why is there a discrepancy?  Can you document it?

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 3]
```diff
@@ -5855,33 +5950,54 @@ fn stack_extend_signer_key() {
...
 
-    let mut reward_set = get_reward_set_entries_at(&mut peer, &latest_block, reward_cycle_ht);
-    assert_eq!(reward_set.len(), 1);
-    let reward_entry = reward_set.pop().unwrap();
-    assert_eq!(
-        PoxAddress::try_from_pox_tuple(false, &pox_addr_val).unwrap(),
-        reward_entry.reward_address
-    );
+    let reward_set = get_reward_set_entries_at(&mut peer, &latest_block, reward_cycle_ht);
+    assert_eq!(reward_set.len(), {
```
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:5954]: Same here and below -- curious why the reward set has a different length in Nakamoto.

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 4]
```diff
@@ -7745,23 +8113,24 @@ fn test_scenario_three() {
         .unwrap();
     assert_eq!(david_aggregate_commit_indexed_err, Value::Int(35));
 
+    let david_index = if use_nakamoto { 3 } else { 2 };
```
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:8116]: Why is this?

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 5]
```diff
@@ -8157,7 +8584,11 @@ fn delegate_stack_increase_err() {
     // test that the reward set is empty
     let reward_cycle_ht = burnchain.reward_cycle_to_block_height(next_reward_cycle);
     let reward_set = get_reward_set_entries_at(&mut peer, &latest_block, reward_cycle_ht);
-    assert!(reward_set.is_empty());
+    if use_nakamoto {
```
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:8587]: Why is this?

**Review by jcnelson — COMMENTED** [review]
I've tested this with #5093 and all of @kantai's changes seem to work :tada: 

My biggest ask is to document why there are discrepancies in behavior in the PoX 4 tests when `nakamoto` is true.

File: stackslib/src/net/relay.rs [hunk 6]
```diff
@@ -840,6 +840,7 @@ impl Relayer {
 
         // NOTE: it's `+ 1` because the first Nakamoto block is built atop the last epoch 2.x
         // tenure, right after the last 2.x sortition
+        // TODO: is this true?
```
[the remark this answers is not in the document]
Comment (jcnelson) [reply stackslib/src/net/relay.rs:888]: I'm not sure. Let's test it.

**Comment by jcnelson** [comment]
> reward_cycle_to_block_height considers the modulo 0 block to be the start of the reward cycle, so I think this is already consistent. nakamoto_first_block_of_cycle, is the "reverse" operation, but it also uses the modulo 0 block height.

Yes, you're right.  I confused those two in my head.

> I'm a little unsure of how this should be handled. This function is only used to verify that pox-4 is the current active PoX contract, and it returns an error otherwise. From 2.5 onwards, this will simply pass.

> If the current burn height is 40, and 2.5 activates at 41, what is the expected behavior here (assuming 20 block cycles)? Should it error?

I think for `getstackers.rs`, we'd want to use `nakamoto_first_block_of_cycle` because that's when the signers' tenure starts.  However, you're also right that it doesn't really matter since the PoX 4 contract has been active for multiple cycles already.

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs (hunk 2 above)
Comment (hstove) [reply stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:5270]: Yep I can add a comment - basically, in 3.0 there is a default signer (separate from the accounts used in the test) to ensure that Nakamoto can progress.

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs (hunk 4 above)
Comment (hstove) [reply stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:8116]: Similar to the other issue - when in Nakamoto, there is a default stacker/signer.

**Comment by jferrant** [comment]
I am not 100% positive where specifically but the very first commit of this branch introduces a database lock that seems to break the end_of_tenure test. I think it might be the change to the thread::scope line about the Peer but I have yet to confirm this.

**Comment by jcnelson** [comment]
This tentatively has my approval but let's find out why there's a failing CI test.

**Comment by jferrant** [comment]
> I am not 100% positive where specifically but the very first commit of this branch introduces a database lock (at least I get a ton of logs about this) that seems to break the end_of_tenure test.

Well I can confirm its not the lock. Its something else. This was a red herring as the database locks being excessively logs occur on develop as well though the test passes.

**Review by jcnelson — APPROVED** [review]
